src_lockbased/lazy_skiplist.ml

```diff
@@ -0,0 +1,360 @@
+(*
+ * RMutex - Reentrant mutexes
+ * Copyright (C) 2008 David Teller, LIFO, Universite d'Orleans
+ *               2011 Edgar Friendly <thelema314@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version,
+ * with the special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *)
+[@@@warning "-32"]
+
+module RMutex = struct
+  type owner =
+    {
+      thread : int;       (**Identity of the latest owner (possibly the current owner)*)
+      mutable depth : int (**Number of times the current owner owns the lock.*)
+    }
+
+  type t =
+    {
+      primitive : Mutex.t; (**A low-level mutex, used to protect access to [ownership]*)
+      wait      : Condition.t; (** a condition to wait on when the lock is locked *)
+      mutable ownership : owner option;
+    }
+
+
+  let create () =
+    {
+      primitive = Mutex.create ();
+      wait      = Condition.create ();
+      ownership = None
+    }
+
+  (**
+     Attempt to acquire the mutex, waiting indefinitely
+   *)
+  let lock m =
+    let id = Thread.id (Thread.self ()) in
```

Comment:
Just noting that this will not work, in general, if you have a scheduler that allows fibers to move from one (domain and) (sys)thread to another (like Domainslib) or allows multiple fibers to run on a single thread (like Eio).  Unfortunately we don't yet have a scheduler agnostic way to get the id of a fiber.

Reply:
Thanks for the review! @polytypic 

I must confess that I find the many names & levels of threads like `systhreads, pthreads, domains, lightweight threads, fibers, etc...` quite confusing. So just to say the obvious thing to confirm that I've understood you correctly:

`Thread.id (Thread.self ())` doesn't work as we would expect under effect-based schedulers because they use fibers which are simulated lightweight threads by the runtime and are different from the pthreads used internally by the `Thread` module. Therefore, if for example we use Eio scheduler, all fibers that call `Thread.id` returns the same one value which is the id of the "main" thread? In the case of using Domainslib which moves fibers between domains, the value of `Thread.id` just depends on which domain it is called from regardless of the work fiber is doing

Reply:
Yes, I think you explained it well. 👍

Also, I haven't really reviewed this PR, yet.  Just a random observation from taking a peek at the code.